src/WebJobs.Script/Rpc/RpcInitializationService.cs

```diff
@@ -68,28 +77,44 @@ internal async Task InitializeRpcServerAsync()
 
         internal Task InitializeChannelsAsync()
         {
-            string workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);
-            if (_environment.IsLinuxAppServiceEnvironment())
+            if (IsLinuxEnvironment())
             {
-                return Task.CompletedTask;
+                InitializeChannelsAsync(LanguageWorkerConstants.LinuxOperatingSystemName, _workerRuntime);
             }
-            if (_environment.IsLinuxContainerEnvironment())
+            else
             {
-                return Task.CompletedTask;
+                InitializeChannelsAsync(LanguageWorkerConstants.WindowsOperatingSystemName, _workerRuntime);
             }
-            if (string.IsNullOrEmpty(workerRuntime) && _environment.IsPlaceholderModeEnabled())
+
+            return Task.CompletedTask;
+        }
+
+        private bool IsLinuxEnvironment()
+        {
+            return _environment.IsLinuxContainerEnvironment() || _environment.IsLinuxAppServiceEnvironment();
+        }
+
+        private Task InitializeChannelsAsync(string os, string workerRuntime)
+        {
+            if (StartInPlaceholderMode(workerRuntime))
             {
-                // Only warm up language workers in placeholder mode in worker runtime is not set
-                return Task.WhenAll(_languages.Select(runtime => _languageWorkerChannelManager.InitializeChannelAsync(runtime)));
+                return Task.WhenAll(_osToLanguagesMapping[os].Select(runtime =>
+                    _languageWorkerChannelManager.InitializeChannelAsync(runtime)));
             }
-            if (_languages.Contains(workerRuntime))
+            if (_osToLanguagesMapping[os].Contains(workerRuntime))
             {
                 return _languageWorkerChannelManager.InitializeChannelAsync(workerRuntime);
             }
             return Task.CompletedTask;
         }
 
+        private bool StartInPlaceholderMode(string workerRuntime)
+        {
+            return string.IsNullOrEmpty(workerRuntime) && _environment.IsPlaceholderModeEnabled();
+        }
+
         // To help with unit tests
-        internal void AddSupportedRuntime(string language) => _languages.Add(language);
+        internal void AddSupportedWindowsRuntime(string language) =>
```

Comment:
It would be a better option to keep this generic, `AddSupportedRuntime` and taking the OS as a parameter